File: ecobertura.ui/test/ecobertura/ui/annotation/AnnotationModelAttacherTest.java
package ecobertura.ui.annotation;

import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import org.eclipse.jface.text.source.IAnnotationModel;
import org.eclipse.jface.text.source.IAnnotationModelExtension;
import org.eclipse.ui.texteditor.IDocumentProvider;
import org.eclipse.ui.texteditor.ITextEditor;
import org.junit.Before;
import org.junit.Test;

// TODO define handler for logger and check warnings
public class AnnotationModelAttacherTest {

	abstract class AnnotationModelImpl implements IAnnotationModel, IAnnotationModelExtension {};
	
	private ITextEditor editor;
	private IDocumentProvider provider;
	private IAnnotationModel invalidModel;
	private IAnnotationModel nonNullModel;
	private IAnnotationModelExtension modelExt;
	
	@Before
	public void setUp() throws Exception {
		editor = mock(ITextEditor.class);
		provider = mock(IDocumentProvider.class);
		invalidModel = mock(IAnnotationModel.class);
		nonNullModel = mock(IAnnotationModel.class);
		modelExt = mock(AnnotationModelImpl.class);
	}

	@Test
	public void testAttachToInvalidProvider() {
		when(editor.getDocumentProvider()).thenReturn(null);
		new AnnotationModelAttacher().attachTo(editor);
//		verify(ilog, times(1)).log(status.capture());
//		assertEquals("no document provider found", status.getValue().getMessage());
	}
	
	@Test
	public void testAttachToInvalidModel() {
		when(editor.getDocumentProvider()).thenReturn(provider);
		when(provider.getAnnotationModel(any())).thenReturn(invalidModel);
		new AnnotationModelAttacher().attachTo(editor);
//		verify(ilog, times(1)).log(status.capture());
//		assertTrue(status.getValue().getMessage().startsWith("unknown annotation model extension type"));
	}
	
	@Test
	public void testAttachTo() {
		when(editor.getDocumentProvider()).thenReturn(provider);
		when(provider.getAnnotationModel(any())).thenReturn((IAnnotationModel) modelExt);
		new AnnotationModelAttacher().attachTo(editor);
//		verify(ilog, times(2)).log(status.capture());
//		assertTrue(status.getValue().getMessage().startsWith("CoverageAnnotationModel attached to"));
	}

	@Test
	public void testAttachToExisting() {
		when(editor.getDocumentProvider()).thenReturn(provider);
		when(provider.getAnnotationModel(any())).thenReturn((IAnnotationModel) modelExt);
		when(modelExt.getAnnotationModel(CoverageAnnotationModel.MODEL_ID)).thenReturn(nonNullModel);
		new AnnotationModelAttacher().attachTo(editor);
//		verify(ilog, never()).log(any(IStatus.class));
	}
}

File: ecobertura.ui/src/test/scala/ecobertura/ui/annotation/AnnotationModelAttacherTest.scala
/*
 * This file is part of eCobertura.
 * 
 * Copyright (c) 2009, 2010 Joachim Hofer
 * All rights reserved.
 *
 * eCobertura is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *  
 * eCobertura is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with eCobertura.  If not, see <http://www.gnu.org/licenses/>.
 */
package ecobertura.ui.annotation

import org.junit.Assert._;
import org.mockito.Matchers._;
import org.mockito.Mockito._;

import java.util.logging._

import org.eclipse.jface.text.source._
import org.eclipse.ui.texteditor._
import org.junit._
import org.mockito.ArgumentCaptor;

class AnnotationModelAttacherTest {
	abstract class AnnotationModelImpl extends IAnnotationModel with IAnnotationModelExtension {}
	
	private var handler: Handler = null
	private var record: ArgumentCaptor[LogRecord] = null
	
	private var editor: ITextEditor = null
	private var provider: IDocumentProvider = null
	private var invalidModel: IAnnotationModel = null
	private var nonNullModel: IAnnotationModel = null
	private var modelExt: IAnnotationModelExtension = null
	
	@Before
	def setUp = {
		registerTraceHandler
		mockEnvironment

		def registerTraceHandler = {
			handler = mock(classOf[Handler])
			Logger.getLogger("ecobertura.ui").addHandler(handler)
			Logger.getLogger("ecobertura.ui").setLevel(Level.ALL)
			record = ArgumentCaptor.forClass(classOf[LogRecord])
		}
		
		def mockEnvironment = {
			editor = mock(classOf[ITextEditor])
			provider = mock(classOf[IDocumentProvider])
			invalidModel = mock(classOf[IAnnotationModel])
			nonNullModel = mock(classOf[IAnnotationModel])
			modelExt = mock(classOf[AnnotationModelImpl])
		}
	}

	@Test
	def testAttachToInvalidProvider = {
		when(editor.getDocumentProvider).thenReturn(null)
		AnnotationModelAttacher.attachTo(editor)
		
		verify(handler).publish(record.capture)
		
		val logged = record.getValue
		assertEquals("unable to attach to editors", logged.getMessage)
		assertEquals(Level.WARNING, logged.getLevel)
		val thrown = logged.getThrown
		assertEquals("no document provider found", thrown.getMessage)
		assertEquals(classOf[NoValidAnnotationModelException], thrown.getClass)
	}
	
	@Test
	def testAttachToInvalidModel = {
		when(editor.getDocumentProvider()).thenReturn(provider)
		when(provider.getAnnotationModel(any())).thenReturn(invalidModel)
		AnnotationModelAttacher.attachTo(editor)

		verify(handler).publish(record.capture)
		
		val logged = record.getValue
		assertEquals("unable to attach to editors", logged.getMessage)
		assertEquals(Level.WARNING, logged.getLevel)
		val thrown = logged.getThrown
		assertTrue(thrown.getMessage.startsWith("unknown annotation model extension type"))
		assertEquals(classOf[NoValidAnnotationModelException], thrown.getClass())
	}
	
	@Test
	def testAttachTo = {
		when(editor.getDocumentProvider()).thenReturn(provider)
		when(provider.getAnnotationModel(any())).thenReturn(modelExt.asInstanceOf[IAnnotationModel])
		AnnotationModelAttacher.attachTo(editor)

		verify(handler, times(2)).publish(record.capture())
		
		val logged = record.getValue
		assertTrue(logged.getMessage.startsWith("CoverageAnnotationModel attached to"))
		assertEquals(Level.FINE, logged.getLevel)
	}
	
	@Test
	def testAttachToExisting = {
		when(editor.getDocumentProvider()).thenReturn(provider)
		when(provider.getAnnotationModel(any())).thenReturn(modelExt.asInstanceOf[IAnnotationModel])
		when(modelExt.getAnnotationModel(CoverageAnnotationModel.MODEL_ID)).thenReturn(nonNullModel)
		AnnotationModelAttacher.attachTo(editor)
		verify(handler, never()).publish(any(classOf[LogRecord]))
	}
}
